File: backend/arceus/core/memory_scope.py
from __future__ import annotations

from arceus.core.hippocampus.hippocampus import Hippocampus
from arceus.core.hippocampus.types import MemoryType, MemoryUnit

_MEMORY_PRIORITY = {
    MemoryType.STATIC: 3,
    MemoryType.DYNAMIC: 2,
    MemoryType.WORKING: 1,
}


class ArceusMemoryScope:
    """Arceus adapter for canonical Hippocampus container naming and retrieval."""

    @staticmethod
    def startup_container(startup_id: str) -> str:
        return f"startup:{startup_id}"

    @staticmethod
    def employee_container(startup_id: str, employee_id: str) -> str:
        return f"startup:{startup_id}:emp:{employee_id}"

    @staticmethod
    def task_container(startup_id: str, task_id: str) -> str:
        return f"startup:{startup_id}:task:{task_id}"

    @staticmethod
    def sub_agent_container(startup_id: str, task_id: str, agent_id: str) -> str:
        return f"startup:{startup_id}:task:{task_id}:sub:{agent_id}"

    async def get_memories_for_agent(
        self,
        hippocampus: Hippocampus,
        query: str,
        startup_id: str,
        employee_id: str,
        task_id: str | None = None,
        include_shared: bool = True,
    ) -> list[MemoryUnit]:
        results: list[MemoryUnit] = []

        if include_shared:
            results.extend(
                await hippocampus.recall(
                    query, self.startup_container(startup_id), include_graph=False
                )
            )

        results.extend(
            await hippocampus.recall(
                query,
                self.employee_container(startup_id, employee_id),
                include_graph=False,
            )
        )

        if task_id:
            results.extend(
                await hippocampus.recall(
                    query,
                    self.task_container(startup_id, task_id),
                    include_graph=False,
                )
            )

        return self._deduplicate_by_priority(results)

    def _deduplicate_by_priority(self, results: list[MemoryUnit]) -> list[MemoryUnit]:
        seen: dict[str, MemoryUnit] = {}

        for memory in results:
            existing = seen.get(memory.content)
            if existing is None:
                seen[memory.content] = memory
                continue
            if _MEMORY_PRIORITY.get(memory.memory_type, 0) > _MEMORY_PRIORITY.get(
                existing.memory_type,
                0,
            ):
                seen[memory.content] = memory

        return list(seen.values())

File: server/src/services/memory-scope.ts
import type { HippocampusBridge, MemoryItem, MemoryListItem } from "./hippocampus-contract.js";
import { logger } from "../middleware/logger.js";

export const MemoryContainers = {
  startup: (startupId: string) => `startup:${startupId}`,
  employee: (startupId: string, employeeId: string) => `startup:${startupId}:emp:${employeeId}`,
  task: (startupId: string, taskId: string) => `startup:${startupId}:task:${taskId}`,
  subAgent: (startupId: string, taskId: string, agentId: string) =>
    `startup:${startupId}:task:${taskId}:sub:${agentId}`,
} as const;

export type MemoryVisibility = "private" | "task_scoped" | "startup_shared" | "board_visible";

const MEMORY_PRIORITY: Record<string, number> = {
  static: 3,
  dynamic: 2,
  working: 1,
};

function logMemoryOp(op: string, agentId: string, extra?: Record<string, unknown>) {
  logger.info({ svc: "memory", op, agentId, ...extra }, "memory scope operation");
}

export class MemoryScopeService {
  constructor(private readonly bridge: HippocampusBridge) {}

  async getMemoriesForAgent(
    agentId: string,
    query: string,
    startupId: string,
    employeeId: string,
    taskId?: string,
    includeShared = true,
    topK = 10,
  ): Promise<MemoryItem[]> {
    const recalls: Array<Promise<{ items: MemoryItem[] }>> = [];

    if (includeShared) {
      recalls.push(this.bridge.recall(agentId, query, MemoryContainers.startup(startupId), topK));
    }
    recalls.push(this.bridge.recall(agentId, query, MemoryContainers.employee(startupId, employeeId), topK));
    if (taskId) {
      recalls.push(this.bridge.recall(agentId, query, MemoryContainers.task(startupId, taskId), topK));
    }

    const settled = await Promise.allSettled(recalls);
    const results: MemoryItem[] = [];

    for (const result of settled) {
      if (result.status === "fulfilled") {
        results.push(...result.value.items);
      }
    }

    logMemoryOp("scoped-recall", agentId, {
      startupId,
      employeeId,
      taskId: taskId ?? null,
      includeShared,
      topK,
      containers: recalls.length,
      results: results.length,
      rejected: settled.filter((item) => item.status === "rejected").length,
    });

    return this.deduplicateByPriority(results);
  }

  async getShareableMemories(
    agentId: string,
    startupId: string,
    visibility: MemoryVisibility[] = ["startup_shared", "board_visible"],
  ): Promise<MemoryListItem[]> {
    const all = await this.bridge.listMemories(agentId, undefined, MemoryContainers.startup(startupId));
    const filtered = all.items.filter((item) => visibility.includes((item.visibility ?? "private") as MemoryVisibility));

    logMemoryOp("shareable-memories", agentId, {
      startupId,
      visibility,
      total: all.total,
      returned: filtered.length,
    });

    return filtered;
  }

  private deduplicateByPriority(items: MemoryItem[]): MemoryItem[] {
    const seen = new Map<string, MemoryItem>();
    for (const item of items) {
      const existing = seen.get(item.content);
      if (!existing) {
        seen.set(item.content, item);
        continue;
      }

      const currentPriority = MEMORY_PRIORITY[item.memory_type ?? ""] ?? 0;
      const existingPriority = MEMORY_PRIORITY[existing.memory_type ?? ""] ?? 0;
      if (currentPriority > existingPriority) {
        seen.set(item.content, item);
      }
    }
    return [...seen.values()];
  }
}
